### ISSUE TITLE
hal deploy apply causes java.lang.NullPointerException on private cluster

### ISSUE THREAD

**Issue by celamb4** [spinnaker/spinnaker#3577]
### Issue Summary:
Switched deployment clusters today... Beforehand this issue was not present. When running hal deploy apply with a clean cluster, the deployment almost completes. When running hal deploy apply again, I receive a nullpointer and only boostraps get deployed.

### Cloud Provider(s):
GKE/k8s

### Environment:
Private GKE cluster
hal deploy to GKE k8s cluster v1.11
external redis
hal --> 1.12.0-20181024113436
spinnaker --> 1.9.5, 1.10.1, 1.10.2

### Feature Area: 
hal

### Description:
Clean slate:
```
> + Deploy spin-clouddriver
>   Success
> + Deploy spin-front50
>   Success
> + Deploy spin-orca
>   Success
> + Deploy spin-deck
>   Success
> + Deploy spin-echo
>   Success
> + Deploy spin-gate
>   Success
> + Deploy spin-igor
>   Success
> + Deploy spin-rosco
>   Success
> Problems in Global:
> ! ERROR Unexpected exception: java.lang.NullPointerException
> 
> - Failed to deploy Spinnaker.
> - 
```

Without DEBUG:
```
> on http://localhost:34839/ui/s cluster in account "myk8scluster": View the kube ui - Deploy spin-front50
>   Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:36107/ui/s cluster in account "myk8scluster": View the kube ui - Apply deployment
>   Failure
> - Deploy spin-clouddriver
>   Failure
> - Deploy spin-front50
>   Failure
> - Deploy spin-orca
>   Failure
> - Deploy spin-deck
>   Failure
> - Deploy spin-echo
>   Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:42467/ui/
> - Deploy spin-gate
>   Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:46815/ui/
> - Deploy spin-igor
>   Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:44355/ui/
> - Deploy spin-rosco
>   Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:44115/ui/
> Problems in Global:
> ! ERROR Unexpected exception: java.lang.NullPointerException
```

With DEBUG enabled:
```
> Problems in Global:
> ! ERROR Unexpected exception: java.lang.NullPointerException
> 
> - Failed to deploy Spinnaker.
> com.netflix.spinnaker.halyard.cli.services.v1.ExpectedDaemonFailureException: Failed to deploy Spinnaker.
> 	at com.netflix.spinnaker.halyard.cli.services.v1.OperationHandler.get(OperationHandler.java:45)
> 	at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.runRemoteAction(AbstractRemoteActionCommand.java:50)
> 	at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.executeThis(AbstractRemoteActionCommand.java:103)
> 	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.safeExecuteThis(NestableCommand.java:201)
> 	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:149)
> 	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:152)
> 	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:152)
> 	at com.netflix.spinnaker.halyard.cli.Main.main(Main.java:46)
> Caused by: java.lang.Exception: Unexpected exception: java.lang.NullPointerException
> 	at com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler.lambda$reduceChildren$0(DaemonTaskHandler.java:59)
> 	at java.util.stream.ReduceOps$1ReducingSink.accept(ReduceOps.java:80)
> 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
> 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
> 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
> 	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
> 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> 	at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:484)
> 	at com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler.reduceChildren(DaemonTaskHandler.java:41)
> 	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:231)
> 	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:56)
> 	at com.netflix.spinnaker.halyard.deploy.services.v1.DeployService.deploy(DeployService.java:287)
> 	at com.netflix.spinnaker.halyard.controllers.v1.DeploymentController.lambda$deploy$20(DeploymentController.java:262)
> 	at com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder.build(DaemonResponse.java:127)
> 	at com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository.lambda$submitTask$1(TaskRepository.java:48)
> 	at java.lang.Thread.run(Thread.java:748)
```

### Steps to Reproduce:
Create private cluster (GKE). 
Nodes to use NAT gateway for outgoing traffic
Authorize connection from haylard VM to master node
Ensure kubectl get namespaces works
add k8s account as deployment target
hal deploy apply

### Additional Details:
For the account used to deploy, I switched from v1 to v2 and then back to v1. Uncertain if this may have caused an issue. 

---

**Comment by celamb4** [comment]
Seems like its unable to get the status of rosco which is running fine,

```
Determining status of spin-rosco: Waiting on deployments to complete
- Deploy spin-clouddriver
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:43601/ui/s cluster in account "myk8scluster": View the kube ui - Deploy spin-front50
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:34079/ui/s cluster in account "myk8scluster": View the kube ui ^ Apply deployment
Determining status of spin-rosco: Waiting on deployments to complete
^ Deploy spin-clouddriver
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:43601/ui/s cluster in account "myk8scluster": View the kube ui ^ Deploy spin-front50
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:34079/ui/s cluster in account "myk8scluster": View the kube ui * Apply deployment
Determining status of spin-rosco: Waiting on deployments to complete
* Deploy spin-clouddriver
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:43601/ui/s cluster in account "myk8scluster": View the kube ui * Deploy spin-front50
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:34079/ui/s cluster in account "myk8scluster": View the kube ui _ Apply deployment
Determining status of spin-rosco: Waiting on deployments to complete
_ Deploy spin-clouddriver
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:43601/ui/s cluster in account "myk8scluster": View the kube ui _ Deploy spin-front50
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:34079/ui/s cluster in account "myk8scluster": View the kube ui - Apply deployment
Failure
- Deploy spin-clouddriver
Failure
- Deploy spin-front50
Failure
- Deploy spin-orca
Failure
! Deploy spin-deck
Interrupted
! Deploy spin-echo
Interrupted
_ Deploy spin-gate! ERROR Unexpected exception: java.lang.NullPointerException
```

**Comment by celamb4** [comment]
Tried updating hal with no luck. Running halyard locally on ubuntu in a VM next to k8s cluster.
```
$ hal -v
1.12.0-20181024113436
```

**Comment by celamb4** [comment]
Tried skipping all services to debug. Ended up with NPE as well.
```
hal deploy apply --exclude-service-names=clouddriver,front50,orca,deck,echo,gate,rosco,igor
+ Preparation complete... deploying Spinnaker
+ Get current deployment
 Success
- Apply deployment
 Failure
Problems in Global:
! ERROR Unexpected exception: java.lang.NullPointerException

- Failed to deploy Spinnaker.
```

Im able to list namespaces. I know hal can deploy services since when its a new deployment (hal deploy clean && hal deploy apply) it creates all services but fails at the end. So dont think its a network or permission issue. Unless hal does something right after bootstrap-orca that throws the NPE

**Comment by celamb4** [comment]
Here are some more logs. Not sure what its attempting to do on localhost. 
```
Connecting to the Kubernetes cluster in account "myk8scluster": View the kube ui on http://localhost:37377/ui/
---> HTTP GET http://localhost:8064/v1/tasks/43e59d99-8356-4fc0-ba9c-e5158ebcc5b9/
---> END HTTP (no body)
<--- HTTP 200 http://localhost:8064/v1/tasks/43e59d99-8356-4fc0-ba9c-e5158ebcc5b9/ (2ms)
X-Application-Context: halyard:test,local:8064
Content-Type: application/json;charset=UTF-8
Transfer-Encoding: chunked
Date: Thu, 08 Nov 2018 15:56:01 GMT
OkHttp-Sent-Millis: 1541692561651
OkHttp-Received-Millis: 1541692561653

{
"name" : "Apply deployment",
"timeout" : 1800000,
"events" : [ {
  "message" : "Building service endpoints",
  "stage" : "Generating all Spinnaker profile files and endpoints",
  "timestamp" : 1541692420099
}, {
  "message" : "Generated 2 profiles and discovered 1 custom profile for clouddriver-bootstrap",
  "stage" : "Generating all Spinnaker profile files and endpoints",
  "timestamp" : 1541692422975
}, {
  "message" : "Generated 3 profiles and discovered 1 custom profile for orca-bootstrap",
  "stage" : "Generating all Spinnaker profile files and endpoints",
  "timestamp" : 1541692423318
}, {
  "message" : "Generated 0 profiles and discovered 0 custom profiles for redis-bootstrap",
  "stage" : "Generating all Spinnaker profile files and endpoints",
  "timestamp" : 1541692423319
}, {
  "message" : "Generated 0 profiles and discovered 0 custom profiles for redis",
  "stage" : "Generating all Spinnaker profile files and endpoints",
  "timestamp" : 1541692423319
}, {
  "message" : "Manually deploying spin-redis-bootstrap",
  "stage" : "Determining status of spin-redis-bootstrap",
  "timestamp" : 1541692423488
}, {
  "message" : "Manually deploying spin-clouddriver-bootstrap",
  "stage" : "Determining status of spin-clouddriver-bootstrap",
  "timestamp" : 1541692446217
}, {
  "message" : "Manually deploying spin-orca-bootstrap",
  "stage" : "Determining status of spin-orca-bootstrap",
  "timestamp" : 1541692495682
}, {
  "message" : "Waiting on deployments to complete",
  "stage" : "Determining status of spin-orca-bootstrap",
  "timestamp" : 1541692539225
}, {
  "message" : "Connected to kubernetes cluster for account myk8scluster on port 37377",
  "stage" : "Connecting to the Kubernetes cluster in account \"myk8scluster\"",
  "timestamp" : 1541692541309
}, {
  "message" : "View the kube ui on http://localhost:37377/ui/",
  "stage" : "Connecting to the Kubernetes cluster in account \"myk8scluster\"",
  "timestamp" : 1541692541309
} ],
"children" : [ ],
"uuid" : "43e59d99-8356-4fc0-ba9c-e5158ebcc5b9",
"timedOut" : false,
"version" : "1.12.0-20181024113436",
"state" : "FAILED",
"response" : {
  "responseBody" : null,
  "problemSet" : {
    "problems" : [ {
      "message" : "Unexpected exception: java.lang.NullPointerException",
      "remediation" : null,
      "options" : null,
      "severity" : "FATAL",
      "location" : "Global"
    } ]
  }
},
"fatalError" : {
  "cause" : null,
  "stackTrace" : [ {
    "methodName" : "reapRoscoServerGroups",
    "fileName" : "DistributedDeployer.java",
    "lineNumber" : 320,
    "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
    "nativeMethod" : false
  }, {
    "methodName" : "deploy",
    "fileName" : "DistributedDeployer.java",
    "lineNumber" : 237,
    "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
    "nativeMethod" : false
  }, {
    "methodName" : "deploy",
    "fileName" : "DistributedDeployer.java",
    "lineNumber" : 56,
    "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
    "nativeMethod" : false
  }, {
    "methodName" : "deploy",
    "fileName" : "DeployService.java",
    "lineNumber" : 287,
    "className" : "com.netflix.spinnaker.halyard.deploy.services.v1.DeployService",
    "nativeMethod" : false
  }, {
    "methodName" : "lambda$deploy$20",
    "fileName" : "DeploymentController.java",
    "lineNumber" : 262,
    "className" : "com.netflix.spinnaker.halyard.controllers.v1.DeploymentController",
    "nativeMethod" : false
  }, {
    "methodName" : "build",
    "fileName" : "DaemonResponse.java",
    "lineNumber" : 127,
    "className" : "com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder",
    "nativeMethod" : false
  }, {
    "methodName" : "lambda$submitTask$1",
    "fileName" : "TaskRepository.java",
    "lineNumber" : 48,
    "className" : "com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository",
    "nativeMethod" : false
  }, {
    "methodName" : "run",
    "fileName" : "Thread.java",
    "lineNumber" : 748,
    "className" : "java.lang.Thread",
    "nativeMethod" : false
  } ],
  "message" : null,
  "localizedMessage" : null,
  "suppressed" : [ ]
},
"runningJobs" : [ ]
- Apply deployment
Failure
Problems in Global:
! ERROR Unexpected exception: java.lang.NullPointerException

- I know everything hasn't been quite right with me, but I can
assure you now, very confidently, that it's going to be alright again.
- Failed to deploy Spinnaker.
com.netflix.spinnaker.halyard.cli.services.v1.ExpectedDaemonFailureException: Failed to deploy Spinnaker.
at com.netflix.spinnaker.halyard.cli.services.v1.OperationHandler.get(OperationHandler.java:45)
at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.runRemoteAction(AbstractRemoteActionCommand.java:50)
at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.executeThis(AbstractRemoteActionCommand.java:103)
at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.safeExecuteThis(NestableCommand.java:201)
at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:149)
at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:152)
at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:152)
at com.netflix.spinnaker.halyard.cli.Main.main(Main.java:46)
Caused by: java.lang.Exception: null
at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.reapRoscoServerGroups(DistributedDeployer.java:320)
at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:237)
at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:56)
at com.netflix.spinnaker.halyard.deploy.services.v1.DeployService.deploy(DeployService.java:287)
at com.netflix.spinnaker.halyard.controllers.v1.DeploymentController.lambda$deploy$20(DeploymentController.java:262)
at com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder.build(DaemonResponse.java:127)
at com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository.lambda$submitTask$1(TaskRepository.java:48)
at java.lang.Thread.run(Thread.java:748)
```

**Comment by celamb4** [comment]
I can confirm the issue is related to GKE private cluster. I created a new cluster and was able to deploy. Not certain what haylards actual requirements are since kubectl can communicate without issues through a NAT Gateway.

**Comment by ryan-mf** [comment]
I have the exact same issue, filed issue within minutes of this one: 3576

Can probably close mine as dupe. Thanks for providing the additional diagnostic data @celamb4! Glad to see it looks to be directly related to running GKE in private cluster mode.

In my case, even though I run into this error, it *does* actually deploy a working Spinnaker stack, I just have to update the services for Deck and Gate (as we always do) to type: LoadBalancer and give it our static public IPs for Spinnaker, and we're back in action.

I also run an "external" Redis to get data (pipeline history) persistence.

*However*, where normally "hal deploy apply" detects our customized Deck and Gate services and leaves them alone while deploying/updating everything else, whatever this issue is, is causing Spinnaker to wipe out my services and replace them with the default ones without type: LoadBalancer and static external IPs.

**Comment by lwander** [comment]
@ryan-mf or @celamb4 have you been able to retry this with the symptom fix merged here: https://github.com/spinnaker/halyard/pull/1101? It's obviously not the complete solution, but allows us to see what the error is.

Similarly, are you able to run `kubectl port-forward -n spinnaker <rosco-pod-id> 8087` and reach rosco's health check endpoint at `localhost:8087/health`?

**Comment by ryan-mf** [comment]
@lwander no we haven't yet, but if it's merged and the updated halyard container is available from the standard registry location, I'll test it today.

**Comment by lwander** [comment]
You will have to try with the `:nightly` tag. 

In the mean time, the second question:

> Similarly, are you able to run kubectl port-forward -n spinnaker <rosco-pod-id> 8087 and reach rosco's health check endpoint at localhost:8087/health

should hopefully point me in the right direction for fixing this

**Comment by ryan-mf** [comment]
Sorry for the delay @lwander, I ran out of time Thursday and was out on Friday.

> Similarly, are you able to run kubectl port-forward -n spinnaker 8087 and reach rosco's health check endpoint at localhost:8087/health

This works using kubectl from my workstation, and I'm able to see "UP" in the response of that URL.

**Comment by ryan-mf** [comment]
Also, we're hoping to test one evening this week with the :nightly tag. Can't do it during the day because everyone here depends on Spinnaker so no daytime maintenance :)

**Comment by lwander** [comment]
Have you had any luck?

I'm still unable to reproduce, so the I'm hoping the reporting we fixed in the above linked PR will point us in the right direction.

**Comment by ryan-mf** [comment]
We have planned Spinnaker maintenance on Thursday night where we'll be testing the :nightly tag to run hal deploy.

**Comment by lwander** [comment]
Thanks!

**Comment by ryan-mf** [comment]
```
* Deploy spin-igor
  Connecting to the Kubernetes cluster in account "xxxxxxxxxx": View the kube ui on http://localhost:37251/ui/
* Deploy spin-rosco
  Connecting to the Kubernetes cluster in account "xxxxxxxxxx": View the kube ui on http://localhost:34597/ui/
Problems in Global:
! ERROR Unexpected exception: java.lang.NullPointerException

- Failed to deploy Spinnaker.
```

Unfortunately the :nightly tag is still suffering this issue it looks like.

**Comment by lwander** [comment]
That's odd, it looks from another issue like something might be up with builds.

Can you share the stack trace? It'll be in `/var/log/spinnaker/halyard/halyard.log`

**Comment by ryan-mf** [comment]
@lwander to answer your questions:
1. We're talking to the cluster from a halyard container running on the same cluster where we deploy spinnaker. Halyard runs in the default namespace, and Spinnaker is deployed to the Spinnaker namespace. This has always worked fine pre-GKE Private Mode.
2. Authentication for the Kubernetes provider is with a kubeconfig file that contains the creds (master user/pass) for the "Corp" K8S cluster (where Spinnaker is deployed) and all of our dev/stage/prod "product" clusters  - all of them are now GKE private mode and Spinnaker is able to deploy our application pipelines to the "private" mode clusters without issue. It only seems to be Halyard itself having a problem deploying.
3. We've made no other material changes since going to GKE Private Mode. We did convert to IP Aliases as well so that our PODs could route to internal IP ranges in our private VPC, but that was done weeks before we want to private mode and caused no observable problems with Halyard deployment.

Maybe there's some issue running Halyard *in* a private mode cluster and deploying Spinnaker (to the Spinnaker namespace) in that same cluster? We don't use a separate bastion host to run Halyard.

Thanks!
-R

**Comment by lwander** [comment]
> Maybe there's some issue running Halyard in a private mode cluster and deploying Spinnaker (to the Spinnaker namespace) in that same cluster?

Good question, let me find out

**Comment by lwander** [comment]
Still having trouble reproducing this -- but someone internally suggested opening firewall rules on the master node for the spinnaker services (roughly 7002, 7003, 808*, and 9000).

**Comment by ryan-mf** [comment]
Hm, we *do* limit access to the master in the K8S cluster using GKE's own settings for specifying an IP whitelist. However, all of the work we're doing, both locally and from the Halyard POD running inside the same K8S cluster that runs Spinnaker, is all in that whitelist.

**Comment by jdstone-loansnap** [comment]
I work under @ryan-mf.  Has there been any movement on this issue?

**Comment by jdstone-loansnap** [comment]
I updated to the latest nightly version of Halyard `1.14.0-20190117020510` and now when I run `hal deploy apply`, I get the following:

```
$ hal deploy apply
+ Get current deployment
  Success
+ Prep deployment
  Success
+ Preparation complete... deploying Spinnaker
+ Get current deployment
  Success
- Apply deployment
  Failure
+ Deploy spin-clouddriver
  Success
+ Deploy spin-front50
  Success
+ Deploy spin-orca
  Success
+ Deploy spin-deck
  Success
+ Deploy spin-echo
  Success
+ Deploy spin-gate
  Success
+ Deploy spin-igor
  Success
+ Deploy spin-rosco
  Success
Problems in Global:
! ERROR Rosco is enabled, and no connection to rosco could be
  established: retrofit.RetrofitError: timeout: no response body

- Failed to deploy Spinnaker.
```

**Comment by jdstone-loansnap** [comment]
I did get the above error, but it looks like the original `NullPointerException` error still exists with the latest nightly build as of Friday, January 18, 2019.

**Comment by lwander** [comment]
Either on the halyard daemon pod, or from another pod in the cluster, can you repeat the steps taken here?

https://github.com/spinnaker/spinnaker/issues/3577#issuecomment-443936157

Clearly this works from outside the cluster, but since you run halyard within the cluster this should show different results.

Also, since you've upgraded halyard, can you share the stack trace? That error looks different from the NPE shown above.

**Comment by jdstone-loansnap** [comment]
@lwander, I ran this command from inside the Halyard pod `kubectl port-forward -n spinnaker spin-rosco-v###-XXXXX 8087` and then ran `kubectl port-forward halyard-dc-v###-XXXXX 8082:8087` from my local computer. I then connected to http://localhost:8082/health from my local computer and received `{"status":"UP"}`.

**Comment by jdstone-loansnap** [comment]
As for the stack trace of the below error, I am no longer able to reproduce that error.  However, I still receive the `NullPointerException` error.
```
ERROR Rosco is enabled, and no connection to rosco could be established: retrofit.RetrofitError: timeout: no response body
```

**Comment by lwander** [comment]
Thanks @jdstone-loansnap -- how about the NPE stack trace as a sanity check for me?

**Comment by jdstone-loansnap** [comment]
You're welcome.  And thank you @lwander for all your help.

I was able to run the command again, this time with the debug parameter, `hal -d deploy apply`.  Hopefully this is what you're looking for.  If it is not, please let me know.
```
* Apply deployment
  Determining status of spin-rosco: Waiting on deployments to complete
* Deploy spin-clouddriver
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:39784/ui/
* Deploy spin-front50
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:45664/ui/
* Deploy spin-orca
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:35313/ui/
* Deploy spin-deck
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:39773/ui/
* Deploy spin-echo
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:43931/ui/
* Deploy spin-gate
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:45272/ui/
* Deploy spin-igor
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:35850/ui/
* Deploy spin-rosco
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:33391/ui/
---> HTTP GET http://localhost:8064/v1/tasks/89ec521b-c2cb-42dc-af8c-3134bc2ce02d/
---> END HTTP (no body)
<--- HTTP 200 http://localhost:8064/v1/tasks/89ec521b-c2cb-42dc-af8c-3134bc2ce02d/ (5ms)
X-Application-Context: halyard:test,local:8064
Content-Type: application/json;charset=UTF-8
Transfer-Encoding: chunked
Date: Fri, 25 Jan 2019 06:49:06 GMT
OkHttp-Sent-Millis: 1548398947008
OkHttp-Received-Millis: 1548398947013

{
  "name" : "Apply deployment",
  "timeout" : 1800000,
  "events" : [ {
    "message" : "Building service endpoints",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398771716
  }, {
    "message" : "Generated 2 profiles and discovered 0 custom profiles for clouddriver-bootstrap",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398774189
  }, {
    "message" : "Generated 2 profiles and discovered 0 custom profiles for clouddriver",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398774356
  }, {
    "message" : "Generated 1 profile and discovered 0 custom profiles for deck",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398774568
  }, {
    "message" : "Generated 2 profiles and discovered 0 custom profiles for echo",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398774721
  }, {
    "message" : "Generated 2 profiles and discovered 0 custom profiles for front50",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398774958
  }, {
    "message" : "Generated 2 profiles and discovered 0 custom profiles for gate",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398775200
  }, {
    "message" : "Generated 2 profiles and discovered 0 custom profiles for igor",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398775343
  }, {
    "message" : "Generated 8 profiles and discovered 0 custom profiles for monitoring-daemon",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398775920
  }, {
    "message" : "Generated 3 profiles and discovered 1 custom profile for orca-bootstrap",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398776366
  }, {
    "message" : "Generated 2 profiles and discovered 1 custom profile for orca",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398776574
  }, {
    "message" : "Generated 0 profiles and discovered 0 custom profiles for redis-bootstrap",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398776575
  }, {
    "message" : "Generated 0 profiles and discovered 0 custom profiles for redis",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398776575
  }, {
    "message" : "Generated 18 profiles and discovered 0 custom profiles for rosco",
    "stage" : "Generating all Spinnaker profile files and endpoints",
    "timestamp" : 1548398777610
  }, {
    "message" : "Manually deploying spin-redis-bootstrap",
    "stage" : "Determining status of spin-redis-bootstrap",
    "timestamp" : 1548398777796
  }, {
    "message" : "Manually deploying spin-clouddriver-bootstrap",
    "stage" : "Determining status of spin-clouddriver-bootstrap",
    "timestamp" : 1548398807538
  }, {
    "message" : "Manually deploying spin-orca-bootstrap",
    "stage" : "Determining status of spin-orca-bootstrap",
    "timestamp" : 1548398878686
  }, {
    "message" : "Waiting on deployments to complete",
    "stage" : "Determining status of spin-rosco",
    "timestamp" : 1548398923820
  } ],
  "children" : [ {
    "name" : "Deploy spin-clouddriver",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 39784",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398924903
    }, {
      "message" : "View the kube ui on http://localhost:39784/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398924903
    } ],
    "children" : [ ],
    "uuid" : "1eeb7534-356d-43c0-ae3f-f219172e74c6",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "FAILED",
    "response" : {
      "responseBody" : null,
      "problemSet" : {
        "problems" : [ {
          "message" : "Unexpected exception: java.lang.NullPointerException",
          "remediation" : null,
          "options" : null,
          "severity" : "FATAL",
          "location" : "Global"
        } ]
      }
    },
    "fatalError" : {
      "cause" : null,
      "stackTrace" : [ {
        "methodName" : "lambda$deploy$4",
        "fileName" : "DistributedDeployer.java",
        "lineNumber" : 218,
        "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
        "nativeMethod" : false
      }, {
        "methodName" : "build",
        "fileName" : "DaemonResponse.java",
        "lineNumber" : 127,
        "className" : "com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder",
        "nativeMethod" : false
      }, {
        "methodName" : "lambda$submitTask$1",
        "fileName" : "TaskRepository.java",
        "lineNumber" : 48,
        "className" : "com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository",
        "nativeMethod" : false
      }, {
        "methodName" : "run",
        "fileName" : "Thread.java",
        "lineNumber" : 748,
        "className" : "java.lang.Thread",
        "nativeMethod" : false
      } ],
      "message" : null,
      "localizedMessage" : null,
      "suppressed" : [ ]
    },
    "runningJobs" : [ ]
  }, {
    "name" : "Deploy spin-front50",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 45664",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925027
    }, {
      "message" : "View the kube ui on http://localhost:45664/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925027
    } ],
    "children" : [ ],
    "uuid" : "1a5c73fb-52c2-49e8-90d0-6c90b51a6fb3",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "RUNNING",
    "response" : null,
    "fatalError" : null,
    "runningJobs" : [ "c05d55d6-f8cf-45d1-b34d-32584ca21f36" ]
  }, {
    "name" : "Deploy spin-orca",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 35313",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925157
    }, {
      "message" : "View the kube ui on http://localhost:35313/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925157
    } ],
    "children" : [ ],
    "uuid" : "380f9a16-55b0-404a-aa6e-de5854e04da9",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "RUNNING",
    "response" : null,
    "fatalError" : null,
    "runningJobs" : [ "fff21526-2430-41ff-9ac4-20f8085f4c89" ]
  }, {
    "name" : "Deploy spin-deck",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 39773",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925296
    }, {
      "message" : "View the kube ui on http://localhost:39773/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925296
    } ],
    "children" : [ ],
    "uuid" : "a7d34d06-3b1a-4b1c-a6b4-45ab817d9d0a",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "INTERRUPTED",
    "response" : {
      "responseBody" : null,
      "problemSet" : {
        "problems" : [ {
          "message" : "Unexpected exception: java.lang.NullPointerException",
          "remediation" : null,
          "options" : null,
          "severity" : "FATAL",
          "location" : "Global"
        } ]
      }
    },
    "fatalError" : {
      "cause" : null,
      "stackTrace" : [ {
        "methodName" : "lambda$deploy$4",
        "fileName" : "DistributedDeployer.java",
        "lineNumber" : 218,
        "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
        "nativeMethod" : false
      }, {
        "methodName" : "build",
        "fileName" : "DaemonResponse.java",
        "lineNumber" : 127,
        "className" : "com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder",
        "nativeMethod" : false
      }, {
        "methodName" : "lambda$submitTask$1",
        "fileName" : "TaskRepository.java",
        "lineNumber" : 48,
        "className" : "com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository",
        "nativeMethod" : false
      }, {
        "methodName" : "run",
        "fileName" : "Thread.java",
        "lineNumber" : 748,
        "className" : "java.lang.Thread",
        "nativeMethod" : false
      } ],
      "message" : null,
      "localizedMessage" : null,
      "suppressed" : [ ]
    },
    "runningJobs" : [ ]
  }, {
    "name" : "Deploy spin-echo",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 43931",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925433
    }, {
      "message" : "View the kube ui on http://localhost:43931/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925433
    } ],
    "children" : [ ],
    "uuid" : "e1d3066e-b1fe-472c-b27d-9da921c397ae",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "RUNNING",
    "response" : null,
    "fatalError" : null,
    "runningJobs" : [ "fb97ebe2-cc31-4cd7-881f-89aa012254b8" ]
  }, {
    "name" : "Deploy spin-gate",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 45272",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925558
    }, {
      "message" : "View the kube ui on http://localhost:45272/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925558
    } ],
    "children" : [ ],
    "uuid" : "0c02732e-cb6f-4986-b9f1-ab0128ccee32",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "RUNNING",
    "response" : null,
    "fatalError" : null,
    "runningJobs" : [ "46378b69-aefd-4141-9912-6de6a6fa533e" ]
  }, {
    "name" : "Deploy spin-igor",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 35850",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925688
    }, {
      "message" : "View the kube ui on http://localhost:35850/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925688
    } ],
    "children" : [ ],
    "uuid" : "f66f80a7-45f0-4d15-a2b3-5bec18d79c42",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "RUNNING",
    "response" : null,
    "fatalError" : null,
    "runningJobs" : [ "287d137a-175d-4bb3-a9f1-1449567ea7cf" ]
  }, {
    "name" : "Deploy spin-rosco",
    "timeout" : 1800000,
    "events" : [ {
      "message" : "Connected to kubernetes cluster for account gke-XXX-us-central1b-XXX on port 33391",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925834
    }, {
      "message" : "View the kube ui on http://localhost:33391/ui/",
      "stage" : "Connecting to the Kubernetes cluster in account \"gke-XXX-us-central1b-XXX\"",
      "timestamp" : 1548398925834
    } ],
    "children" : [ ],
    "uuid" : "12feed64-be4a-4f7c-a5d3-84e82812f551",
    "timedOut" : false,
    "version" : "1.14.0-20190117020510",
    "state" : "RUNNING",
    "response" : null,
    "fatalError" : null,
    "runningJobs" : [ "2cc2cbec-18d5-4bb3-946d-8fdb04a5ff4c" ]
  } ],
  "uuid" : "89ec521b-c2cb-42dc-af8c-3134bc2ce02d",
  "timedOut" : false,
  "version" : "1.14.0-20190117020510",
  "state" : "FAILED",
  "response" : {
    "responseBody" : null,
    "problemSet" : {
      "problems" : [ {
        "message" : "Unexpected exception: java.lang.NullPointerException",
        "remediation" : null,
        "options" : null,
        "severity" : "FATAL",
        "location" : "Global"
      } ]
    }
  },
  "fatalError" : {
    "cause" : null,
    "stackTrace" : [ {
      "methodName" : "lambda$reduceChildren$0",
      "fileName" : "DaemonTaskHandler.java",
      "lineNumber" : 59,
      "className" : "com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler",
      "nativeMethod" : false
    }, {
      "methodName" : "accept",
      "fileName" : "ReduceOps.java",
      "lineNumber" : 80,
      "className" : "java.util.stream.ReduceOps$1ReducingSink",
      "nativeMethod" : false
    }, {
      "methodName" : "forEachRemaining",
      "fileName" : "ArrayList.java",
      "lineNumber" : 1382,
      "className" : "java.util.ArrayList$ArrayListSpliterator",
      "nativeMethod" : false
    }, {
      "methodName" : "copyInto",
      "fileName" : "AbstractPipeline.java",
      "lineNumber" : 481,
      "className" : "java.util.stream.AbstractPipeline",
      "nativeMethod" : false
    }, {
      "methodName" : "wrapAndCopyInto",
      "fileName" : "AbstractPipeline.java",
      "lineNumber" : 471,
      "className" : "java.util.stream.AbstractPipeline",
      "nativeMethod" : false
    }, {
      "methodName" : "evaluateSequential",
      "fileName" : "ReduceOps.java",
      "lineNumber" : 708,
      "className" : "java.util.stream.ReduceOps$ReduceOp",
      "nativeMethod" : false
    }, {
      "methodName" : "evaluate",
      "fileName" : "AbstractPipeline.java",
      "lineNumber" : 234,
      "className" : "java.util.stream.AbstractPipeline",
      "nativeMethod" : false
    }, {
      "methodName" : "reduce",
      "fileName" : "ReferencePipeline.java",
      "lineNumber" : 484,
      "className" : "java.util.stream.ReferencePipeline",
      "nativeMethod" : false
    }, {
      "methodName" : "reduceChildren",
      "fileName" : "DaemonTaskHandler.java",
      "lineNumber" : 41,
      "className" : "com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler",
      "nativeMethod" : false
    }, {
      "methodName" : "deploy",
      "fileName" : "DistributedDeployer.java",
      "lineNumber" : 232,
      "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
      "nativeMethod" : false
    }, {
      "methodName" : "deploy",
      "fileName" : "DistributedDeployer.java",
      "lineNumber" : 56,
      "className" : "com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer",
      "nativeMethod" : false
    }, {
      "methodName" : "deploy",
      "fileName" : "DeployService.java",
      "lineNumber" : 288,
      "className" : "com.netflix.spinnaker.halyard.deploy.services.v1.DeployService",
      "nativeMethod" : false
    }, {
      "methodName" : "lambda$deploy$14",
      "fileName" : "DeploymentController.java",
      "lineNumber" : 210,
      "className" : "com.netflix.spinnaker.halyard.controllers.v1.DeploymentController",
      "nativeMethod" : false
    }, {
      "methodName" : "build",
      "fileName" : "DaemonResponse.java",
      "lineNumber" : 127,
      "className" : "com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder",
      "nativeMethod" : false
    }, {
      "methodName" : "lambda$submitTask$1",
      "fileName" : "TaskRepository.java",
      "lineNumber" : 48,
      "className" : "com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository",
      "nativeMethod" : false
    }, {
      "methodName" : "run",
      "fileName" : "Thread.java",
      "lineNumber" : 748,
      "className" : "java.lang.Thread",
      "nativeMethod" : false
- Apply deployment
  Failure
- Deploy spin-clouddriver
  Failure
* Deploy spin-front50
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:45664/ui/
* Deploy spin-orca
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:35313/ui/
! Deploy spin-deck
  Interrupted
* Deploy spin-echo
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:43931/ui/
* Deploy spin-gate
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:45272/ui/
* Deploy spin-igor
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:35850/ui/
* Deploy spin-rosco
  Connecting to the Kubernetes cluster in account "gke-XXX-us-central1b-XXX": View the kube ui on http://localhost:33391/ui/
Problems in Global:
! ERROR Unexpected exception: java.lang.NullPointerException

- Failed to deploy Spinnaker.
com.netflix.spinnaker.halyard.cli.services.v1.ExpectedDaemonFailureException: Failed to deploy Spinnaker.
  at com.netflix.spinnaker.halyard.cli.services.v1.OperationHandler.get(OperationHandler.java:45)
  at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.runRemoteAction(AbstractRemoteActionCommand.java:50)
  at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.executeThis(AbstractRemoteActionCommand.java:103)
  at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.safeExecuteThis(NestableCommand.java:201)
  at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:149)
  at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:152)
  at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:152)
  at com.netflix.spinnaker.halyard.cli.Main.main(Main.java:46)
Caused by: java.lang.Exception: Unexpected exception: java.lang.NullPointerException
  at com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler.lambda$reduceChildren$0(DaemonTaskHandler.java:59)
  at java.util.stream.ReduceOps$1ReducingSink.accept(ReduceOps.java:80)
  at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
  at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
  at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
  at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
  at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
  at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:484)
  at com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler.reduceChildren(DaemonTaskHandler.java:41)
  at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:232)
  at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:56)
  at com.netflix.spinnaker.halyard.deploy.services.v1.DeployService.deploy(DeployService.java:288)
  at com.netflix.spinnaker.halyard.controllers.v1.DeploymentController.lambda$deploy$14(DeploymentController.java:210)
  at com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder.build(DaemonResponse.java:127)
  at com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository.lambda$submitTask$1(TaskRepository.java:48)
  at java.lang.Thread.run(Thread.java:748)
```

**Comment by jdstone-loansnap** [comment]
Hi @lwander.  Some more info.  Two days ago (January 28th), from the Halyard container, I ran the following command to enable Slack notifications, `echo $TOKEN_FROM_SLACK | hal config notification slack edit --bot-name spinnakerbot --token` and then ran `hal deploy apply` to deploy/apply the changes. This is where I encountered the same error I received previously.
```
$ hal deploy apply
+ Get current deployment
  Success
+ Prep deployment
  Success
+ Preparation complete... deploying Spinnaker
+ Get current deployment
  Success
- Apply deployment
  Failure
+ Deploy spin-clouddriver
  Success
+ Deploy spin-front50
  Success
+ Deploy spin-orca
  Success
+ Deploy spin-deck
  Success
+ Deploy spin-echo
  Success
+ Deploy spin-gate
  Success
+ Deploy spin-igor
  Success
+ Deploy spin-rosco
  Success
Problems in Global:
! ERROR Rosco is enabled, and no connection to rosco could be
  established: retrofit.RetrofitError: timeout: no response body

- Failed to deploy Spinnaker.
```
But this time I have the stack trace you wanted.  I've attached that here.  The log starts a few lines before I ran `hal deploy apply` (and after I ran the above Slack notification command).  The first error that I see appears on line 2010.  Hope this helps and let me know if there is anything else I can provide to you or test on your behalf.
[halyard_log_20190128.log](https://github.com/spinnaker/spinnaker/files/2815617/halyard_log_20190128.log)

**Comment by jdstone-loansnap** [comment]
@lwander, I ran this command from inside the Halyard pod `kubectl proxy` and then ran `kubectl port-forward halyard-dc-v###-XXXXX 8001` from my local computer. I then connected to http://localhost:8001/api/v1/namespaces/spinnaker/services/spin-rosco:8087/proxy/status/all from my local computer and received the following.  Is this what you wanted me to perform?
```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "services \"spin-rosco\" not found",
  "reason": "NotFound",
  "details": {
    "name": "spin-rosco",
    "kind": "services"
  },
  "code": 404
}
```

**Comment by jdstone-loansnap** [comment]
I get the following when navigating to http://localhost:8001.
```json
{
  "paths": [
    "/api",
    "/api/v1",
    "/apis",
    "/apis/",
    "/apis/admissionregistration.k8s.io",
    "/apis/admissionregistration.k8s.io/v1beta1",
    "/apis/apiextensions.k8s.io",
    "/apis/apiextensions.k8s.io/v1beta1",
    "/apis/apiregistration.k8s.io",
    "/apis/apiregistration.k8s.io/v1",
    "/apis/apiregistration.k8s.io/v1beta1",
    "/apis/apps",
    "/apis/apps/v1",
    "/apis/apps/v1beta1",
    "/apis/apps/v1beta2",
    "/apis/authentication.k8s.io",
    "/apis/authentication.k8s.io/v1",
    "/apis/authentication.k8s.io/v1beta1",
    "/apis/authorization.k8s.io",
    "/apis/authorization.k8s.io/v1",
    "/apis/authorization.k8s.io/v1beta1",
    "/apis/autoscaling",
    "/apis/autoscaling/v1",
    "/apis/autoscaling/v2beta1",
    "/apis/batch",
    "/apis/batch/v1",
    "/apis/batch/v1beta1",
    "/apis/certificates.k8s.io",
    "/apis/certificates.k8s.io/v1beta1",
    "/apis/cloud.google.com",
    "/apis/cloud.google.com/v1beta1",
    "/apis/extensions",
    "/apis/extensions/v1beta1",
    "/apis/metrics.k8s.io",
    "/apis/metrics.k8s.io/v1beta1",
    "/apis/networking.k8s.io",
    "/apis/networking.k8s.io/v1",
    "/apis/policy",
    "/apis/policy/v1beta1",
    "/apis/rbac.authorization.k8s.io",
    "/apis/rbac.authorization.k8s.io/v1",
    "/apis/rbac.authorization.k8s.io/v1beta1",
    "/apis/scalingpolicy.kope.io",
    "/apis/scalingpolicy.kope.io/v1alpha1",
    "/apis/storage.k8s.io",
    "/apis/storage.k8s.io/v1",
    "/apis/storage.k8s.io/v1beta1",
    "/healthz",
    "/healthz/autoregister-completion",
    "/healthz/etcd",
    "/healthz/ping",
    "/healthz/poststarthook/apiservice-openapi-controller",
    "/healthz/poststarthook/apiservice-registration-controller",
    "/healthz/poststarthook/apiservice-status-available-controller",
    "/healthz/poststarthook/bootstrap-controller",
    "/healthz/poststarthook/ca-registration",
    "/healthz/poststarthook/generic-apiserver-start-informers",
    "/healthz/poststarthook/kube-apiserver-autoregistration",
    "/healthz/poststarthook/rbac/bootstrap-roles",
    "/healthz/poststarthook/start-apiextensions-controllers",
    "/healthz/poststarthook/start-apiextensions-informers",
    "/healthz/poststarthook/start-kube-aggregator-informers",
    "/healthz/poststarthook/start-kube-apiserver-informers",
    "/logs",
    "/metrics",
    "/openapi/v2",
    "/swagger-2.0.0.json",
    "/swagger-2.0.0.pb-v1",
    "/swagger-2.0.0.pb-v1.gz",
    "/swagger.json",
    "/swaggerapi",
    "/version"
  ]
}
```

**Comment by lwander** [comment]
Is Rosco running?

I'm transferring to another team this week, I'm going to hand this off to
@jtk54

**Comment by jdstone-loansnap** [comment]
Thank you @lwander for all your help!

Hi @jtk54.  We run Spinnaker in a GCP Kubernetes cluster.  The Rosco container is running and I was able to log into the container and run `ps aux | grep -i rosco`, so it looks like it's running.  Is there a better way to confirm?

**Comment by mvillafuertem** [comment]
I have same error, any solution? 

`2019-02-20 18:00:14.728 ERROR 2960 --- [     Thread-398] c.n.s.halyard.core.DaemonResponse        : Unknown exception encountered: 

java.lang.NullPointerException: null
	at com.netflix.spinnaker.halyard.config.model.v1.providers.google.GoogleAccount.getNamedAccountCredentials(GoogleAccount.java:57) ~[halyard-config-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.spinnaker.v1.service.distributed.google.GoogleProviderUtils.getCompute(GoogleProviderUtils.java:410) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.spinnaker.v1.service.distributed.google.GoogleDistributedService.getRunningServiceDetails(GoogleDistributedService.java:524) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:195) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:56) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.services.v1.DeployService.deploy(DeployService.java:288) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.controllers.v1.DeploymentController.lambda$deploy$14(DeploymentController.java:210) ~[halyard-web-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder.build(DaemonResponse.java:127) ~[halyard-core-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository.lambda$submitTask$1(TaskRepository.java:48) ~[halyard-core-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_201]

2019-02-20 18:00:14.729  WARN 2960 --- [     Thread-398] c.n.s.h.core.tasks.v1.TaskRepository     : Task [Apply deployment] (13dfbedb-5357-43c3-aaab-5bfe8aef84bf) - RUNNING failed with unexpected reason: 

java.lang.NullPointerException: null
	at com.netflix.spinnaker.halyard.config.model.v1.providers.google.GoogleAccount.getNamedAccountCredentials(GoogleAccount.java:57) ~[halyard-config-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.spinnaker.v1.service.distributed.google.GoogleProviderUtils.getCompute(GoogleProviderUtils.java:410) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.spinnaker.v1.service.distributed.google.GoogleDistributedService.getRunningServiceDetails(GoogleDistributedService.java:524) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:195) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:56) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.deploy.services.v1.DeployService.deploy(DeployService.java:288) ~[halyard-deploy-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.controllers.v1.DeploymentController.lambda$deploy$14(DeploymentController.java:210) ~[halyard-web-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder.build(DaemonResponse.java:127) ~[halyard-core-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository.lambda$submitTask$1(TaskRepository.java:48) ~[halyard-core-1.15.0-20190206103304.jar:1.15.0-20190206103304]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_201]`

**Comment by jtk54** [comment]
@mvillafuertem @jdstone-loansnap The quickest way I know to check that k8s/gke thinks the services are running is to do a `kubectl get po -n spinnaker` and see which pods are spun up. Judging by the hal output in the first post, it seems like the k8s ops are actually completing but the halyard process is dying for the NPE reasons (sorry if I'm reiterating, coming up to speed on this bug).

Can you attempt to connect to the running Spinnaker after hal fails and let me know what happens?

**Comment by jtk54** [comment]
Seems suspicious: https://github.com/spinnaker/halyard/commit/82923b1f7903078255915fd47eb1ebcab27e05d8#diff-5908b80ecfa50ae837086b491106d426. I'll see what's going on in that change and get back to you.

**Comment by jtk54** [comment]
Which hal version are you folks running?

**Comment by jtk54** [comment]
I think switching back and forth fro k8s v1 -> v2 -> v1 as the deployer account might have caused issues, I think switching to v2 might solve these issues.

@jdstone-loansnap are you also seeing the GoogleAccount NPE or are we crossing two different hal problems?

**Comment by exequielrafaela** [comment]
@celamb4 We were having the same error and after changing external redis to the default redis deployment via values.yml config the helm deployment was successful. Have you already tried this scenario? 

`redis:
  enabled: true`

**Comment by jdstone-loansnap** [comment]
@jtk54,
My apologies for not getting back to you sooner -- I've been extremely busy.

- Immediately after `hal deploy apply` fails, I was able to successfully connect to Spinnaker with no issues.
- We're running `hal version: 1.14.0-20190117020510`.
> @jdstone-loansnap are you also seeing the GoogleAccount NPE or are we crossing two different hal problems?
- We are **_not_** seeing the GoogleAccount NPE issue.
> I think switching back and forth fro k8s v1 -> v2 -> v1 as the deployer account might have caused issues, I think switching to v2 might solve these issues.
- Can you please clarify what you mean by "switching back and forth from K8s v1 -> v2 -> v1?  We've been using the v2 provider for some time now.  Unless I'm misunderstanding what you're asking?

**Comment by billimek** [comment]
We were experiencing a NPE deploying spinnaker via halyard to a GKE cluster using the v3 kubernetes provider.

While it wasn't a private cluster, I want to share out findings in case anyone else reading this issue is in a similar situation:

We are also using SSL and found [this thread on slack](https://spinnakerteam.slack.com/archives/C2W6RE0N4/p1549907840010800) which states,

> Hey @slavakl we had the same issue earlier, caused by the missing flag "ssl-certificate-passphrase". This causes a null pointer in the application when no value is provided, for us this occurred because our certificate doesn't have a passphrase.

> There are 2 solutions to this issue:
> - Create an SSL Cert with a passphrase and include the flag, with the corresponding value.
> - If your cert does not have a passphrase, use the flag with a random string and funny enough it works

... indeed, our NPE problem went away when we set the cert password via,

```shell
hal config security ui ssl edit --ssl-certificate-passphrase 
<when prompted enter the password>
```

... We further automated this via expect,

```shell
expect -c "spawn hal config security ui ssl edit --ssl-certificate-passphrase; sleep 1; expect -exact \"The passphrase needed to unlock your SSL certificate. This will be provided to Apache on startup.: \"; send -- \"${PASSWORD}\r\"; expect eof"
```

**Comment by jdstone-loansnap** [comment]
Hi @jtk54,

Did you see my latest post (reply to your post)?  Is there more information that you need or has a resolution been found?  Should I update to the latest stable version of Halyard or the latest nightly build?

**Comment by jtk54** [comment]
Sorry for the delay. I'm referring to 

> For the account used to deploy, I switched from v1 to v2 and then back to v1. Uncertain if this may have caused an issue.

in https://github.com/spinnaker/spinnaker/issues/3577#issue-378624330. I suspect this might have caused some of the initial issues. I believe the latest Halyard release is 1.18.x, so it may be worth your while to upgrade.

**Comment by jdstone-loansnap** [comment]
Thank you @jtk54.

Oh, in our case, we did **not** switch from v1 to v2 and then back to v1 -- we made the switch from v1 to v2 and did **not** switch back -- once we made the switch to v2, we stuck with it.

I'll update to Halyard v1.18.x and report back if the issue still continues.

**Comment by jtk54** [comment]
:+1:

**Comment by jdstone-loansnap** [comment]
Hello @jtk54,

I updated Halyard to version `1.18.0-20190325172326` and while re-deploying Spinnaker (running `hal deploy apply`, I got the following error, again:
```
$ hal deploy apply
+ Get current deployment
  Success
+ Prep deployment
  Success
+ Preparation complete... deploying Spinnaker
+ Get current deployment
  Success
- Apply deployment
  Failure
+ Deploy spin-clouddriver
  Success
+ Deploy spin-front50
  Success
+ Deploy spin-orca
  Success
+ Deploy spin-deck
  Success
+ Deploy spin-echo
  Success
+ Deploy spin-gate
  Success
+ Deploy spin-igor
  Success
+ Deploy spin-rosco
  Success
Problems in Global:
! ERROR Rosco is enabled, and no connection to rosco could be
  established: retrofit.RetrofitError: timeout: no response body

- Failed to deploy Spinnaker.
```
I ran what @lwander suggested [here](https://github.com/spinnaker/spinnaker/issues/3577#issuecomment-443936157): I ran this command from inside the Halyard pod `kubectl port-forward -n spinnaker spin-rosco-v###-XXXXX 8087` and then ran `kubectl port-forward halyard-dc-v###-XXXXX 8082:8087` from my local computer. I then connected to http://localhost:8082/health from my local computer and received `{"status":"UP"}`.

I also did again what @lwander suggested [here](https://github.com/spinnaker/spinnaker/issues/3577#issuecomment-459363549): I ran this command from inside the Halyard pod `kubectl proxy` and then ran `kubectl port-forward halyard-dc-v###-XXXXX 8001` from my local computer. From my local computer, I wasn't able to navigate to http://localhost:8001/api/v1/namespaces/spinnaker/services/spin-rosco:8087/proxy/status/all, it just sat there forever and eventually timed out and gave me the following message (Google Chrome Developer tools gave me a `503 Service Unavailable` status code):
```
Error: 'dial tcp 10.16.0.139:8087: getsockopt: connection timed out'
Trying to reach: 'http://10.16.0.139:8087/status/all'
```
(`10.16.0.139` is the internal IP of the spin-rosco pod)

But I was able to navigate to http://localhost:8001/api/v1/namespaces/spinnaker/services/spin-rosco and got the following:
```
{
  "kind": "Service",
  "apiVersion": "v1",
  "metadata": {
    "name": "spin-rosco",
    "namespace": "spinnaker",
    "selfLink": "/api/v1/namespaces/spinnaker/services/spin-rosco",
    "uid": "ff97278f-5758-11e9-bafb-4201ac100003",
    "resourceVersion": "19423859",
    "creationTimestamp": "2019-04-05T04:12:16Z",
    "labels": {
      "app": "spin",
      "stack": "rosco"
    }
  },
  "spec": {
    "ports": [
      {
        "name": "http",
        "protocol": "TCP",
        "port": 8087,
        "targetPort": 8087
      },
      {
        "name": "monitoring",
        "protocol": "TCP",
        "port": 8008,
        "targetPort": 8008
      }
    ],
    "selector": {
      "load-balancer-spin-rosco": "true"
    },
    "clusterIP": "10.16.30.157",
    "type": "ClusterIP",
    "sessionAffinity": "None"
  },
  "status": {
    "loadBalancer": {
      
    }
  }
}
```
When I navigate to http://localhost:8001/api/v1/namespaces/spinnaker/services/spin-rosco:8087, I get the following:
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "services \"spin-rosco:8087\" not found",
  "reason": "NotFound",
  "details": {
    "name": "spin-rosco:8087",
    "kind": "services"
  },
  "code": 404
}
```

Please let me know if there is anything else you need.

**Comment by jtk54** [comment]
This feels like a networking misconfiguration/miscommunication either in halyard's query of the k8s master node. Is there a particular reason why the port `8087` needs to be in-lined in the request?

What does 'http://localhost:8001/api/v1/namespaces/spinnaker/services/spin-rosco/proxy/status/all' give you?

**Comment by jdstone-loansnap** [comment]
I'm not sure why the port `8087` needs to be in-lined in the request -- that's the link that @lwander gave me to try.

I opened up the tunnel again
>...I ran this command from inside the Halyard pod `kubectl proxy` and then ran `kubectl port-forward halyard-dc-v###-XXXXX 8001` from my local computer.

and navigated to the link you provided (http://localhost:8001/api/v1/namespaces/spinnaker/services/spin-rosco/proxy/status/all), and got the following.
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "services \"spin-rosco\" not found",
  "reason": "NotFound",
  "details": {
    "name": "spin-rosco",
    "kind": "services"
  },
  "code": 404
}
```

Keep in mind that this Halyard problem started occurring when we moved to a `Private GKE cluster`.

**Comment by jtk54** [comment]
Genuinely having a tough time figuring out what is happening here. In order to debug this properly I think you'd need to run the Halyard code manually with logging/tracing in place to get the information required to solve this and also likely look at the k8s master node logs as well. It's not clear that private gke cluster support for Halyard is something we should prioritize over other efforts in the k8s (and larger spinnaker) space. I've added this to the k8s v2 backlog to track so we eventually solve this.

**Comment by jdstone-loansnap** [comment]
Thank you @jtk54.  If I have time, I will definitely get you those logs you mentioned above.  After I retrieve the logs for you, I'm going to rebuild our cluster with non-Private mode disabled.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'to-be-closed' and hasn't been updated  in 45 days, so we are closing it. You can always reopen this issue if needed.

**Comment by Sytzeandreae** [comment]
I'm reopening this issue, we're running into it as we speak.

We're also trying to deploy Spinnaker into a private GKE cluster, and are getting the following error message:
```
Problems in Global:
! ERROR Unexpected exception: java.lang.NullPointerException

- Failed to deploy Spinnaker.
com.netflix.spinnaker.halyard.cli.services.v1.ExpectedDaemonFailureException: Failed to deploy Spinnaker.
	at com.netflix.spinnaker.halyard.cli.services.v1.OperationHandler.get(OperationHandler.java:44)
	at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.runRemoteAction(AbstractRemoteActionCommand.java:49)
	at com.netflix.spinnaker.halyard.cli.command.v1.AbstractRemoteActionCommand.executeThis(AbstractRemoteActionCommand.java:102)
	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.safeExecuteThis(NestableCommand.java:222)
	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:172)
	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:175)
	at com.netflix.spinnaker.halyard.cli.command.v1.NestableCommand.execute(NestableCommand.java:175)
	at com.netflix.spinnaker.halyard.cli.Main.main(Main.java:46)
Caused by: java.lang.Exception: Unexpected exception: java.lang.NullPointerException
	at com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler.lambda$reduceChildren$0(DaemonTaskHandler.java:61)
	at java.util.stream.ReduceOps$1ReducingSink.accept(ReduceOps.java:80)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:551)
	at com.netflix.spinnaker.halyard.core.tasks.v1.DaemonTaskHandler.reduceChildren(DaemonTaskHandler.java:41)
	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:262)
	at com.netflix.spinnaker.halyard.deploy.deployment.v1.DistributedDeployer.deploy(DistributedDeployer.java:55)
	at com.netflix.spinnaker.halyard.deploy.services.v1.DeployService.deploy(DeployService.java:308)
	at com.netflix.spinnaker.halyard.controllers.v1.DeploymentController.lambda$deploy$14(DeploymentController.java:228)
	at com.netflix.spinnaker.halyard.core.DaemonResponse$StaticRequestBuilder.build(DaemonResponse.java:123)
	at com.netflix.spinnaker.halyard.core.tasks.v1.TaskRepository.lambda$submitTask$1(TaskRepository.java:48)
	at java.lang.Thread.run(Thread.java:748)
```

Hal version: 1.23.2-20190904152725
GKE version: 1.13.7-gke.24

**Comment by SeanTurner026** [comment]
For anyone who ends up here, I had the same issue as  Sytzeandreae, and was able to sort myself out by uninstalling and reinstalling the hal CLI. For some reason the cli wasn't communicating with the daemon correctly. Make sure you back up your `~/.hal` directory first as the uninstall script nukes this directory.
